### PR TITLE
Instantiate argument type vars before implicit search

### PR DESCRIPTION
**Pull request by EugeneFlesselle**

### CODE REVIEW

**Comment by odersky** [comment]
This looks actually quite promising. The one failure in libretto is probably due to the fact that a type variable is not instantiated to a singleton type. So we might have had a constraint
```
A <: String
```
and proceeded to instantiate to`String` but implicit search would have added the stronger constraint `A <: "ping"`.
Just guessing. We should try to minimize the failure to better judge it.

**Comment by EugeneFlesselle** [comment]
> This looks actually quite promising. The one failure in libretto is probably due to the fact that a type variable is not instantiated to a singleton type. So we might have had a constraint `A <: String` and proceeded to instantiate to`String` but implicit search would have added the stronger constraint `A <: "ping"`. Just guessing. We should try to minimize the failure to better judge it.

That is indeed exactly what happens ! 

Here is the minimisation I got to:
```scala

case class Val[T](t: T)

trait Ev[T]

given Ev[String] = ???
given Ev["ping"] = ???
given[T: Ev]: Ev[Val[T]] = ???

def f[T: Ev](v: T): Unit = ???

def g[U](u: U)(using U =:= Val["ping"]): Unit = ???

def main =

  val v /*: Val3[String] */ = Val("ping")

  f(v)           // before changes ✅; after changes ✅
  f(Val("ping")) // before changes ❌; after changes ✅

  g(v)           // before changes ❌; after changes ❌
  g(Val("ping")) // before changes ✅; after changes ❌
```

The instantiation of `T` in `Val[T]("ping")` is widened from `("ping" : String)` to `String` 
https://github.com/lampepfl/dotty/blob/c88c0fe4103ef8fc2e96e0d33368c92036467a61/compiler/src/dotty/tools/dotc/core/ConstraintHandling.scala#L705-L708

**Review by smarter — APPROVED** [review]
Otherwise LGTM. The breakage in Libretto is unfortunate (/cc @TomasMikula) but having played with that code I haven't found an alternative that avoided that. It's possible something akin to https://github.com/scala/improvement-proposals/pull/48 will allow for APIs that can control inference better in the future.

File: compiler/src/dotty/tools/dotc/typer/Inferencing.scala [hunk 1]
```diff
@@ -392,8 +392,10 @@ object Inferencing {
    *    - The result expression `e` of a block `{s1; .. sn; e}`.
    */
   def tvarsInParams(tree: Tree, locked: TypeVars)(using Context): List[TypeVar] = {
-    @tailrec def boundVars(tree: Tree, acc: List[TypeVar]): List[TypeVar] = tree match {
-      case Apply(fn, _) => boundVars(fn, acc)
+    def boundVars(tree: Tree, acc: List[TypeVar]): List[TypeVar] = tree match {
+      case Apply(fn, args) =>
```
[the remark this answers is not in the document]
Comment (EugeneFlesselle) [reply compiler/src/dotty/tools/dotc/typer/Inferencing.scala:396]: Indeed !

**Comment by TomasMikula** [comment]
Thanks for the heads up, @smarter. The breakage from this change is rather minor. I have added a definition

```scala
  transparent inline def constVal[A]: Done -⚬ Val[A] =
    constVal(scala.compiletime.constValue[A])
```

so that the diff to fix the broken code is then just changing `constVal("ping")` to `constVal["ping"]`.

Though before I even got _this_ breakage, I had to resolve a couple of different breakages (one of them reported as #19570). The thing is, the community build is using an old version of Libretto. Last time I tried to update (#17044), the community build did not support Java 9 required by Libretto. Would be good to get it resolved so that both you and me get breakage notifications earlier.

**Comment by WojciechMazur** [comment]
There is one more failure in the Open CB found in @ichoran /kse3 - [build logs](https://github.com/VirtusLab/community-build3/actions/runs/8265031821/job/22612284835) 
I believe it is an actual improvement, so I'm not creating a dedicated issue.

```scala
class Anon[A](val value: A)
class Mu[A]
object Mu:
  final class MuAny[A](v: A) extends Mu[A]

trait Typed[A]
def typed[A]: Typed[A] = ???
class Labeled[A](v: A):
  def ====[B](t: Typed[B])(using B =:= A): Unit = {}

@main def Test =
  Labeled(Anon(new Mu.MuAny("foo"))) ==== typed[Anon[Mu[String]]]
  
  // After typer:
  // In 3.4.0
  // new Labeled[Anon[Mu[String]]](
  //   new Anon[Mu[String]](new Mu.MuAny[String]("cod"))).====[
  //   Anon[Mu[String]]](typed[Anon[Mu[String]]])(<:<.refl[Anon[Mu[String]]])
  // }

  // In 3.4.1-RC1 and 3.4.2-RC1-nightly
  // new Labeled[Anon[Mu.MuAny[String]]](
  //   new Anon[Mu.MuAny[String]](new Mu.MuAny[String]("cod"))).====[
  //   Anon[Mu[String]]](typed[Anon[Mu[String]]])(
  //   /* missing */summon[Anon[Mu[String]] =:= Anon[Mu.MuAny[String]]])

```

**Comment by soronpo** [comment]
This PR caused the regression #20053

**Comment by Ichoran** [comment]
Wait, you added kse3 to the CB?!  That's lovely!  Um!  I'll be a little less cavalier about pushing nonsense into it then! :laughing: 

I agree that kse3 is the one who is wrong here: the types are supposed to match exactly, but don't.  So the new behavior is an improvement because it catches the bug in how the test is written! :+1:

**Comment by Ichoran** [comment]
@WojciechMazur - Sorry for not noticing your message initially!  You're correct that this was (from my perspective) a bug in my test suite that 3.4.1 correctly caught, and I've now fixed it in main.

**Pull request merged**